akka-typed/src/main/scala/akka/typed/internal/DeathWatch.scala

```diff
@@ -41,15 +41,25 @@ private[typed] trait DeathWatch[T] {
 
   type ARImpl = ActorRefImpl[Nothing]
 
-  private var watching = Set.empty[ARImpl]
+  private var watching = Map.empty[ARImpl, Option[T]]
   private var watchedBy = Set.empty[ARImpl]
 
   final def watch[U](_a: ActorRef[U]): Unit = {
     val a = _a.sorry
     if (a != self && !watching.contains(a)) {
       maintainAddressTerminatedSubscription(a) {
         a.sendSystem(Watch(a, self))
-        watching += a
+        watching = watching.updated(a, None)
+      }
+    }
+  }
+
+  final def watchWith[U](_a: ActorRef[U], msg: T): Unit = {
+    val a = _a.sorry
```

Comment:
Ugh, I know this was copy-pasta, but could we not have the variable names `_a` and `a` but ones that mean something?